osquery/events/windows/etw/etw_publisher_processes.cpp

```diff
@@ -0,0 +1,570 @@
+/**
+ * Copyright (c) 2014-present, The osquery authors
+ *
+ * This source code is licensed as defined by the LICENSE file found in the
+ * root directory of this source tree.
+ *
+ * SPDX-License-Identifier: (Apache-2.0 OR GPL-2.0-only)
+ */
+
+#include <ctime>
+
+#include <osquery/core/flags.h>
+#include <osquery/events/windows/etw/etw_publisher_processes.h>
+#include <osquery/utils/conversions/windows/strings.h>
+#include <osquery/utils/map_take.h>
+
+namespace osquery {
+
+FLAG(bool,
+     enable_etw_process_events,
+     false,
+     "Enables the etw_process_events publisher");
+
+// ETW Event publisher registration into the Osquery pub-sub framework
+REGISTER_ETW_PUBLISHER(EtwPublisherProcesses, kEtwProcessPublisherName.c_str());
+
+// Publisher constructor
+EtwPublisherProcesses::EtwPublisherProcesses()
+    : EtwPublisherBase(kEtwProcessPublisherName) {
+  initializeHardVolumeConversions();
+};
+
+// There are multiple ETW providers being setup here. Events arriving from
+// these providers will be aggregated in the post-processing phase.
+Status EtwPublisherProcesses::setUp() {
+  if (!FLAGS_enable_etw_process_events) {
+    return Status::failure(kEtwProcessPublisherName +
+                           " publisher disabled via configuration.");
+  }
+  // ETW Initialization logic
+  const EtwProviderConfig::EtwBitmask processStartStopKeyword = 0x10;
+
+  // Userspace ETW Provider configuration
+  EtwProviderConfig userEtwProviderConfig;
+  userEtwProviderConfig.setName("Microsoft-Windows-Kernel-Process");
+  userEtwProviderConfig.setAnyBitmask(processStartStopKeyword);
+  userEtwProviderConfig.setPreProcessor(GetPreProcessorCallback());
+  userEtwProviderConfig.setPostProcessor(GetPostProcessorCallback());
+  userEtwProviderConfig.addEventTypeToHandle(EtwEventType::ProcessStart);
+  userEtwProviderConfig.addEventTypeToHandle(EtwEventType::ProcessStop);
+
+  // Adding the provider to the ETW Engine
+  Status userProviderAddStatus = EtwEngine().addProvider(userEtwProviderConfig);
+  if (!userProviderAddStatus.ok()) {
+    return userProviderAddStatus;
+  }
+
+  // Kernelspace ETW Provider configuration
+  EtwProviderConfig kernelProviderCfg;
+  kernelProviderCfg.setKernelProviderType(
+      EtwProviderConfig::EtwKernelProviderType::Process);
+  kernelProviderCfg.setPreProcessor(GetPreProcessorCallback());
+  kernelProviderCfg.setPostProcessor(GetPostProcessorCallback());
+  kernelProviderCfg.addEventTypeToHandle(EtwEventType::ProcessStart);
+  kernelProviderCfg.addEventTypeToHandle(EtwEventType::ProcessStop);
+
+  // Adding the provider to the ETW Engine
+  Status kernelProviderAddStatus = EtwEngine().addProvider(kernelProviderCfg);
+  if (!kernelProviderAddStatus.ok()) {
+    return kernelProviderAddStatus;
+  }
+
+  return Status::success();
+}
+
+// Callback to perform pre-processing logic
+void EtwPublisherProcesses::ProviderPreProcessor(
```

Comment:
Just fixed the casing on the function. The `providerPreProcessor()` and `providerPostProcessor()` member functions are callbacks used during the ETW processing pipeline. The `providerPreProcessor()` callback returns void due to a limitation of the ETW callback interface exposed by the OS. This callback is passed to the OS API, and it gets called each time a new ETW event is generated for the registered provider - more info [here](https://learn.microsoft.com/en-us/windows/win32/api/evntrace/nc-evntrace-pevent_record_callback). I've added a bit of logging when it was possible, and I'm also now printing out an error in case anything popped up on the generic callback wrapper [here](https://github.com/osquery/osquery/pull/7821/files/099f97af86532158565421fe0c64b5582fa1c75f#diff-184586cb21c53bb75810aa81456264c90850ef195fb0641237700fdf42686c75R26). Good suggestion!